subprojects/functional/src/main/java/org/gradle/internal/lazy/NonAtomicLazy.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.lazy;
+
+import javax.annotation.Nullable;
+import java.util.function.Supplier;
+
+class NonAtomicLazy<T> implements Lazy<T> {
```

Comment:
💭 The key fact of this class is that the `supplier` can compute the value more than once depending on how many threads attempt to `get()` simultaneously.  `Non-atomic` does not suggest repeated computations to me - it merely highlights that the supplier may be running the computation when another thread observes it and this other thread may see an intermediate state.  But it won't!  Seeing the result of a partial computation isn't a problem, since the computation will be properly re-run from the beginning in this scenario

Or to be more-specific, this isn't a problem _if the computation is idempotent_.

1) I think the name should suggest this restriction more directly.  Maybe `RerunableLazy`?  `RepeatableLazy`?  `EventuallyTrulyLazyLazy`?  `IdempotentLazy`?  I don't think any of these are great names, but I think there must be something better than `NonAtomic` that could suggest the mechanism used here better.

2) In every case except a computation that returns `null`, couldn't we verify that the computation _is in fact repeatable_ when it is run multiple times by checking against the value already present in `value`?

```java
if (s != null) {
  T t = s.get();
  if (t != null && value != null && !t.equals(value)) {
      throw new RuntimeException("Supplier not idempotent, repeated calculautions yielded different values: " + t + " and " + value);
  }
  value = t:
  supplier = null;
  return t;
}
```

This relies on equality being implemented for whatever the return type is, but we could state this as a requirement of the class's constructor.  It could help catch some potentially really nasty bugs where the value varies if computed multiple times.

Reply:
Thank you both for the feedback. I agree it's more like `AtomicReference` specially after the `publish single value` semantics.  I'm satisfied with `Lazy.atomic()`.